File: crates/beet_action/src/scripting/dynamic/world_shim.js
(() => {
	// every call in flight, by id, so a reply settles exactly the promise that
	// asked for it. Nothing here is reachable from the script: the map and the
	// counter live in this closure.
	const pending = new Map();
	let next = 0;

	const call = (op) =>
		new Promise((resolve, reject) => {
			const id = next++;
			pending.set(id, { resolve, reject });
			globalThis.__world_send({ id, ...op });
		});

	// the host's half: one reply per call, settling or rejecting it. An `err`
	// rejects with a real `Error`, so a refused write is catchable at the call
	// site rather than a silent no-op.
	globalThis.__world_reply = (reply) => {
		const settle = pending.get(reply.id);
		if (settle === undefined) return;
		pending.delete(reply.id);
		if (reply.status === "ok") settle.resolve(reply.value);
		else settle.reject(new Error(reply.message));
	};

	// an entity is an opaque string token, so `String(entity)` is what the host
	// parses back; an entity handle stringifies to the same token, so either
	// form may be passed anywhere an entity is taken. `undefined` becomes `null`
	// because JSON has no undefined.
	const id = (entity) => String(entity);
	const json = (value) => (value === undefined ? null : value);

	globalThis.world = {
		get: (entity, component) =>
			call({ op: "get", entity: id(entity), component }),
		entities: (component) => call({ op: "entities", component }),
		schema: (component) => call({ op: "schema", component }),
		spawn: (components) =>
			call({ op: "spawn", components: components || {} }),
		insert: (entity, component, value) =>
			call({
				op: "insert",
				entity: id(entity),
				component,
				value: json(value),
			}),
		remove: (entity, component) =>
			call({ op: "remove", entity: id(entity), component }),
		despawn: (entity) => call({ op: "despawn", entity: id(entity) }),
		get_field: (entity, path) =>
			call({ op: "get_field", entity: id(entity), path }),
		set_field: (entity, path, value) =>
			call({
				op: "set_field",
				entity: id(entity),
				path,
				value: json(value),
			}),
		entity: (entity) => handle(entity),
	};

	// one entity's half of the `world` API, with the entity already supplied.
	// `world.entity(id)` is how a script holds onto one; an event script is
	// handed its own as `target`.
	const handle = (entity) => ({
		id: id(entity),
		toString: () => id(entity),
		get: (component) => globalThis.world.get(entity, component),
		insert: (component, value) =>
			globalThis.world.insert(entity, component, value),
		remove: (component) => globalThis.world.remove(entity, component),
		despawn: () => globalThis.world.despawn(entity),
		get_field: (path) => globalThis.world.get_field(entity, path),
		set_field: (path, value) =>
			globalThis.world.set_field(entity, path, value),
	});
})();

File: crates/beet_action/src/scripting/dynamic/world_shim.rs
//! The `world` global, one source shared by every backend.

/// The `world` API a world-bridged script sees, installed before its source
/// runs.
///
/// Where the two `console` shims mirror each other by convention, this one is
/// literally the same string on every backend: it names one host hook per
/// direction, and each host installs its half the way it can. So there is no
/// per-backend contract to drift.
///
/// - `__world_send(call)` — the host's inbound hook, taking one [`WorldCall`]
///   as a plain object. The embedded engine binds a host function; a host-realm
///   runner wraps it in a `ScriptEvent`.
/// - `__world_reply(reply)` — the shim's own outbound hook, which the host
///   calls with one [`WorldReply`] to settle the promise its `id` names.
///
/// Every `world` method returns a promise the reply settles, so an `await` is a
/// real operation against the live world at the moment it runs: a script reads
/// its own writes, and a refused call rejects at the call site.
///
/// `world.entity(id)` narrows the same API to one entity, the JS face of an
/// [`AsyncEntity`]: `get`/`insert`/`remove`/`despawn` with the entity already
/// supplied, plus the `get_field`/`set_field`/`with_field` document helpers. An
/// event script is handed its own event target as `target`.
///
/// `with_field(path, func)` is the read-apply-write pair composed in the shim,
/// since a closure has no wire form and so could never be a host operation. It
/// is why `target.with_field('count', count => count + 1)` is the whole counter:
/// the read that would otherwise force an `await` happens inside the helper.
///
/// [`AsyncEntity`]: beet_core::prelude::AsyncEntity
///
/// [`WorldCall`]: crate::prelude::WorldCall
/// [`WorldReply`]: crate::prelude::WorldReply
pub(crate) const WORLD_SHIM: &str = include_str!("world_shim.js");

#[cfg(test)]
mod test {
	use super::WORLD_SHIM;
	use beet_core::prelude::*;

	/// The shim names exactly the two hooks every host promises to honour, one
	/// per direction.
	#[beet_core::test]
	fn the_shim_names_its_host_hooks() {
		WORLD_SHIM.xpect_contains("__world_send");
		WORLD_SHIM.xpect_contains("__world_reply");
	}

	/// The entity handle offers every operation an entity can be the subject
	/// of, so `world.entity(id)` and `target` are never a narrower API than the
	/// `world` calls they stand in for.
	#[beet_core::test]
	fn the_handle_offers_every_entity_operation() {
		for method in [
			"get",
			"insert",
			"remove",
			"despawn",
			"get_field",
			"set_field",
			"with_field",
		] {
			WORLD_SHIM.xpect_contains(&format!("{method}: ("));
		}
	}

	/// Every operation the `world` API offers is present in the one shared
	/// source, so no backend can be missing one.
	#[beet_core::test]
	fn the_shim_offers_every_operation() {
		for op in [
			"get",
			"entities",
			"schema",
			"spawn",
			"insert",
			"remove",
			"despawn",
			"get_field",
			"set_field",
		] {
			WORLD_SHIM.xpect_contains(&format!("op: \"{op}\""));
		}
	}
}
